### DISCUSSION TITLE
[Bug] Storefront API does not return up-to-date and consistent responses after a product is added

### DISCUSSION THREAD

**Discussion by wecessary** [Shopify/storefront-api-feedback#191] [Help] [answered]
I am using Next JS 13.3.0 to build a custom storefront. 

After a product is added my store, the Storefront API (2023-04) behaves strangely when I do the following:

1. I add a product using Shopify admin. The product is then automatically added to a collection called "all-products"
2. I verify the product is live by using the admin dashboard and Shopify GraphiQL app. The product is indeed live and can be seen in the collection.
3. I run the following query to grab all product handles,
 ```
query GetProducts($after: String) {
    collection(handle: "all-products") {
      products(first: 250, after: $after) {
        edges {
          node {
            handle
          }
        }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
  }
```
4. The query does not always return edges of the same length. For demonstration, I have run the same query twice in useEffect as such:

```
import { GraphQLClient } from "graphql-request";

  useEffect(() => {
    async function foo() {
      const data1 = await new GraphQLClient(STOREFRONT_ENDPOINT as string, {
        headers: {
          "X-Shopify-Storefront-Access-Token": process.env
            .NEXT_PUBLIC_STOREFRONT_TOKEN as string,
        },
      }).request(GetProducts, {
        after: undefined,
      });
      const data2 = await new GraphQLClient(STOREFRONT_ENDPOINT as string, {
        headers: {
          "X-Shopify-Storefront-Access-Token": process.env
            .NEXT_PUBLIC_STOREFRONT_TOKEN as string,
        },
      }).request(GetProducts, {
        after: undefined,
      });

      console.log("data1", data1?.collection?.products.edges);
      console.log("data2", data2?.collection?.products.edges);
    }
    foo();
  });
```
In the screenshot you can see, most of the times (each time = me hitting refresh on the browser), it returns 38 products, which is incorrect. There should be 39 products, only the circled instance is correct.
![Screenshot 2023-06-17 at 18 03 26](https://github.com/Shopify/storefront-api-feedback/assets/92378671/b3725983-a022-4a00-94ed-77180073e843)

5.  Just to give more context,  I actually use Apollo client in my Next Js app. Initially I thought this issue had to do with caching with the Apollo client so I tested this out with [graphql-request v6.1.0](https://github.com/jasonkuhrt/graphql-request), which is a lightweight graphql library and does no caching. I even made sure I use a new instance of the client to fetch to make sure it has nothing to do with the client.

7.  This is an issue for me to use `getStaticProps` and `getStaticPaths` to generate a product page at build time. As in, if I can't grab a product's handle, I can't generate a url in `getStaticPaths`, and so on. At the moment I am using `fallback:true` in `getStaticPaths`, so that the product page is built once at run time if the handle is not grabbed at build time. But I would still like to get to the bottom of this, as I think this is potentially a Storefront API bug?  

8. Thanks in advance.

**Comment by johncraigcole** [comment]
Hey @wecessary 👋 . It would be best to take this through proper Shopify support channels to ensure you're authenticated on the store in question and hopefully gain context on what might be happening here.

For the support team, please perform the test again and share the following details.

- The date and time the new product was created, and ideally the product id that resulted.
- The `x-request-id` from a subsequent Storefront API query that worked.
- The `x-request-id` from a subsequent Storefront API query that returned the wrong data.

**Reply by ocastx** [reply]
Hey @johncraigcole, thank you for the feedback on @wecessary's reported Bug. We're are experiencing the same strange behavior, (but with Metafields) since about a week.

We already submitted two support requests (one through Partner support, one from a development store directly), but did receive no definitive answer. We included the Request ID as mentioned. Our support request number is 39366722. Maybe the team could have another look at it? Appreciate the support!

**Reply by wecessary** [reply]
Hey  @johncraigcole, thank you for the reply 😄. 

I performed the test again by creating this product:

- `createdAt` : `2023-06-23T10:04:05Z`
- `id`: `gid://shopify/Product/8360790950181`

After adding this product, the store should have 42 products. 

- A storefront API query that **did not work** had x-request-id: "no-uuid-1"
- A storefront API query that **worked** had x-request-id: "no-uuid-2"
![Screenshot 2023-06-23 at 11 22 00](https://github.com/Shopify/storefront-api-feedback/assets/92378671/e9d44a86-7fa3-4740-9141-0109bea0d4aa)

❗ **Please allow me to give more context.** 

1. Initially I was trying to reproduce the bug by calling the following function:
```
async function foo() {
    const uuid1 = uuidv4(); // just some function that creates uuid. See https://github.com/uuidjs/uuid
    const uuid2 = uuidv4(); 
    const data1 = await new GraphQLClient(STOREFRONT_ENDPOINT as string, {
      headers: {
        "X-Shopify-Storefront-Access-Token": process.env
          .NEXT_PUBLIC_STOREFRONT_TOKEN as string,
      },
    }).request(GetProducts, {
      after: undefined,
      // "X-Request-Id": uuid1,
    });
    const data2 = await new GraphQLClient(STOREFRONT_ENDPOINT as string, {
      headers: {
        "X-Shopify-Storefront-Access-Token": process.env
          .NEXT_PUBLIC_STOREFRONT_TOKEN as string,
      },
    }).request(GetProducts, {
      after: undefined,
      // "X-Request-Id": uuid2,
    });

    console.log("data1", uuid1, data1.collection?.products.edges);
    console.log("data2", uuid2, data2?.collection?.products.edges);
  }
```
2. This actually did not manage to reproduce the bug. I called this function many times and it consistently gave me the **_correct_** query response. See screenshot below.
![Screenshot 2023-06-23 at 11 27 42](https://github.com/Shopify/storefront-api-feedback/assets/92378671/d7b93bd0-9bc0-4d14-92ed-bfdf8f451d5e)
3. I was kind of not surprised by this. I remembered that when I first ran into this issue. If I changed the arguments in the query, the query will then give me the correct response. For example:
```
query GetProducts($after: String) {
    collection(handle: "all-products") {
      products(first: 250, after: $after) { // changing the number to anything else would yield a correct response on next query
        ///...rest of the query
       }
```
4. So then I called the same query again, but without providing `x-request-id`. Immediately, I got two incorrect responses. NOTE: even though it console.logged uuids, no uuids were provided in the request header.
![Screenshot 2023-06-23 at 11 33 58](https://github.com/Shopify/storefront-api-feedback/assets/92378671/7e4ba5ec-42da-4895-846d-85ce286a8306)

5. After this happened, I called the same function `foo`, but manually typed uuid1 and uuid2 as "no-uuid-1" and "no-uuid-2" and reproduced the bug.

I hope I used x-request-id in the right way because I actually haven't used it before...

The issue seems to be that if you send the exact same query again, you will get an out-dated response, but if you change the query even just a little bit, you will get an up-to-date response?

Hope this is helpful 🙏 !!

**Reply by johncraigcole** [reply]
Hey @wecessary 👋 The `x-request-id` is in the response headers of the Storefront API request. I'm assuming the GraphQLClient package you're using would provide access to the response headers where you'd extract the values. With that mind, if you wouldn't mind could you provide the `x-request-id` from the response headers of the SFAPI request, for one that returns the correct quantity and the one that doesn't that would be perfect.

**Reply by wecessary** [reply]
Hey @johncraigcole, thanks for looking at this quickly. My apologies, I should've thought of looking at the network tab! 

I've run the test again by adding a new product.

The newly added product:
```
      "id": "gid://shopify/Product/8361264251173",
      "title": "Foo glass bowl",
      "handle": "foo-glass-bowl",
      "createdAt": "2023-06-23T17:17:17Z"
```
There should now be 43 products.

I ran the same query four times and the second time was correct. 

The `x-request-id` are the following, following the order of the screenshot below:

1. 00c2c7ea-f28e-47b0-863a-6de5c529846f
2. 74ce3766-c408-4319-a58e-4cb996a60e48 // <--- this is the one that returns the correct quantity
3. 154f8b02-9130-4f38-bf6e-694387348f80
4. dfd9f6d1-ba65-4f8d-b8a6-5734ea7de52c 


![Screenshot 2023-06-23 at 18 20 24](https://github.com/Shopify/storefront-api-feedback/assets/92378671/4186d0c3-2c9e-4130-8b1d-92cbf144717f)


Sorry about my last reply and thanks again!

**Reply by johncraigcole** [reply]
Hey @wecessary thanks for the updated data and wanted to share an update.

I was curious if the query needs to happen on every render of your react component? Right now `useEffect` has no dependencies. Using an empty array for the second parameter would ensure the fetch only happens on initial render. This may or may not be what you want but just wanted to mention it. I'm also assuming doing the two queries is just to help illustrate the issue and this isn't what you would have in production where you send the same query twice. 

Could we try this again with the single query assuming that's what the production code will look like? I'd also be curious if for example the fetch happened from a user action, like a button click would you see the same thing.

**Reply by johncraigcole** [reply]
Hey @ocastx can you share the context on when the translated fields were updated and then when you tried to query for them. Also, like wecessary's example could you share the request id's with the correct response and the ones with not.

**Comment by johncraigcole** [comment, marked as answer]
Hey @wecessary @ocastx chatting with some folks internally, and we did have a caching problem in one of our regions in Europe. Both your requests were in the same region. We've pushed a fix for the issue so if both of you could test again just to verify that you're no longer seeing the issue that would be great.

If the issue is still appearing please provide the time in UTC the update was made to a resource like a product, along with the `x-request-id` of the Storefront API request that has stale data.

**Reply by ocastx** [reply]
@johncraigcole we tested the issue intensively today and didn't see any caching problems anymore. Yesterday we still experienced them, so the fix seems to be working for us. Thank you for your fast help on this topic! Really appreciate it 🙏

**Reply by wecessary** [reply]
@johncraigcole It looks all good on my end too. Thanks very much for dealing with it! 🥳 

Sorry I missed your last comment - yes the useEffect was just for illustration. In product I use a static site generator (Next Js), so the query is called once only a build time to build a catalogue of products. But because of the issue it meant that the catalogue would sometimes include the latest product and sometimes doesn't.

**Comment by twyststar** [comment]
We are having issues with metafields cached on the storefront - we change them and changes do not reflect. This has taken our install time from about 5 minutes to about an hour just waiting on metafield updates. 
There is also significant lag in the storefront showing updates to theme files - liquid/css. 
Seems to have started right about the same time as the Editions introduction. 

Is this related?

**Comment by peter-visualsquares** [comment]
I encounter similar issues recently, storefront API getCollectionByHandle is returning outdated collection data. For example, sometime after turning a product to active after 5 min, the product still wasn't included in the collection data of the getCollectionByHandle api.

**Comment by mathur-prerit** [comment]
In ASIA region we are also facing same bug when we are trying to fetch  getCollectionById. We have observed that all outdated requests contains a header `x-cache` with value as _hit,server_

**Comment by vamshiaruru32** [comment]
Hi everyone, just to confirm that I am facing this issue in India still.  And it is not just cached too, it gives inconsistent responses. Some times it gives upto date data, sometimes it gives old data. I am seeing this issue mostly after updating a collection image.

These are the resposne headers if it helps:

```
  "X-Cache": [
    "hit, server"
  ],
  "X-Content-Type-Options": [
    "nosniff"
  ],
  "X-Dc": [
    "gcp-asia-south1,gcp-asia-southeast1,gcp-asia-southeast1"
  ],
  ```

Setting no cache header doesn't seem to help :/ 

Any help is very appreciated, as some of our flows are breaking because of this!

**Reply by gulam-quinn** [reply]
same here, from india, facing same issue. we are only fetching metaobejcts and stale data is breaking a lot of things.

**Comment by jonasivy** [comment]
Hi everyone, did we already fixed the issue?  Saw some comments that it's still occurring on their end.

**Comment by gulam-quinn** [comment]
Just confirming, yes this is happening for us as well. We are fetching metaobejcts using storefrontapi and sometime updated data takes couple of days to reflect on stores. Just so you know this is happening only on couple of stores for some reason. On other stores it takes 5-10 minutes, which is reasonable but taking more than a day-two is not ideal. Request is served by `gcp-asia-south1,gcp-asia-southeast1,gcp-asia-southeast1` servers.

I have created new discussion here https://github.com/Shopify/storefront-api-feedback/discussions/199

**Comment by johncraigcole** [comment]
Hey Everyone 👋.

The Storefront API read queries are served from our global edge layer. For performance, read queries are cached, but should be invalidated when a given resource is updated. When something is updated in the Admin or via Admin API, this kicks off a bunch of processes to update data used by the edge layer and invalidate any cache keys. Depending on system load, this entire process could take between 5-30 seconds for our entire global edge network to have the given update. 

Seeing `hit,server` would indicate that particular query found a match in our edge cache and the cached response is returned to the API consumer. You can not send any additional request headers to tell Shopify not to cache the Storefront API read query response.

For this type of investigation it would be best to contact support so we can make sure your authenticated on the store in question and be able to discuss details about the store. If this is reproducible, including the following in the support request would be helpful.

- The date and time in UTC the resource in question was updated along with the type (product, collection) and id.
- The `x-request-id` from the response headers of the Storefront API request where incorrect data was returned.

We have a limited log retention window so sharing the above with support on the day the issue is reproduced would be beneficial. If you already have a ticket with support, please share the ticket number and I will try and help out here.

**Comment by takamario** [comment]
Let me ask a question since we encountered a similar behavior, though it's a slightly different topic.

Does the cache behavior also occur in Storefront customer API?

In our case, after changing the customer's email, we recreate a token using `customerAccessTokenCreate` mutation with new email.
Strangely, it returns the same token as before. However, `customer` query still returns the old customer information.

Alternatively, in another case, calling `customerAccessTokenRenew` mutation returns the same token as before, but  `customer` query returns the updated customer information.
(Perhaps it returns the updated data because the caches expired while calling `customerAccessTokenRenew`)

**Comment by swayanshu-apbr** [comment]
I am facing a similar issue in cartUpdate mutation. 
x-request-id
600c001b-39ff-4786-a611-fe08d8a30fcf-1737457627
